Snakefile

```diff
@@ -233,48 +225,115 @@ rule files:
+def input_by_name(name, metadata=False, segment=False):
+    assert (metadata and not segment) or (segment and not metadata), "Workflow bug"
+    info = next((c for c in [*config['inputs'], *config.get('additional_inputs', [])] if c['name']==name))
+    return info.get('metadata', None) if metadata else _segment_input(info, segment)
 
-def input_sequences(wildcards):
-    if S3_SRC:
-        return f"data/{S3_SRC['name']}/sequences_{wildcards.segment}.fasta",
-    elif LOCAL_INGEST:
-        return f"ingest/{LOCAL_INGEST}/results/sequences_{wildcards.segment}.fasta"
-    raise Exception() # already caught by `sanity_check_config` above, this is just being cautious
+def collect_inputs(metadata=None, segment=None, augur_merge=False):
+    """
+    This function is pure - subsequent calls will return the same results.
+    """
+    assert (metadata and not segment) or (segment and not metadata), "Workflow bug"
+
+    inputs = []
+    for source in [*config['inputs'], *config.get('additional_inputs', [])]:
+        name = source['name']
+
+        address = source.get('metadata', None) if metadata else _segment_input(source, segment)
+        if address is None:
+            continue # inputs can define (e.g.) only metadata, or only sequences for some segments etc
+
+        # addresses may be a remote filepath or a local file
+        if address.startswith('s3://'):
+            download_path = f"data/{source['name']}/metadata.tsv" \
+                if metadata \
+                else f"data/{source['name']}/sequences_{segment}.fasta"
+            inputs.append((name, download_path))
+            continue
+        elif address.startswith(r'http[s]:\/\/'):
+            raise InvalidConfigError("Workflow cannot yet handle HTTP[S] inputs")
+        inputs.append((name, resolve_config_path(address, {'segment':segment})))
+
+    if not inputs:
+        raise InvalidConfigError("No inputs provided with defined metadata")
+
+    if augur_merge:
+        return " ".join([f"{x[0]}={x[1]}" for x in inputs])
+    return [x[1] for x in inputs]
+
+rule merge_metadata:
+    """
+    This rule should only be invoked if there are multiple defined metadata inputs
+    (config.inputs + config.additional_inputs)
+    """
+    input:
+        metadata = collect_inputs(metadata=True)
+    params:
+        metadata = collect_inputs(metadata=True, augur_merge=True)
+    output:
+        metadata = "results/metadata_merged.tsv"
+    shell:
+        r"""
+        augur merge \
+            --metadata {params.metadata} \
+            --source-columns 'input_{{NAME}}' \
+            --output-metadata {output.metadata}
+        """
 
+rule merge_sequences:
+    """
+    Placeholder rule while we want for `augur merge` to support sequences
+    """
+    input:
+        metadata = lambda w: collect_inputs(segment=w.segment)
+    output:
+        metadata = "results/sequences_merged_{segment}.fasta"
+    shell:
+        r"""
+        cat {input.metadata} > {output.metadata}
```

Comment:
Thanks @victorlin! I used `cat` with the note that this is a "Placeholder rule while we want for `augur merge` to support sequences". If I switch to your `seqkit` approach now what would the eventual differences be cf. `augur merge sequences`?

Reply:
Assuming that `augur merge --sequences` will be more or less the current implementation in the open PR, the `cat` approach will result in different behavior in the case of duplicates within or across files. The `seqkit` approach is nearly identical to what `augur merge --sequences` does under the hood. The difference is that `augur merge --sequences` comes with added validation ([e.g.](https://github.com/nextstrain/augur/blob/81b22d7949b147c06565c95727d7ff221ebe957f/tests/functional/merge/cram/merge-sequences.t#L54)).

Another difference is that `augur merge` can warn on mismatches between metadata and sequences in the same command ([e.g.](https://github.com/nextstrain/augur/blob/81b22d7949b147c06565c95727d7ff221ebe957f/tests/functional/merge/cram/merge-metadata-and-sequences.t#L154)). This is not handled by either of the `cat` and `seqkit` approaches.

Summary: I think the `seqkit` approach is technically more appropriate as a "placeholder rule" for `augur merge --sequences`, but I wouldn't consider this blocking if `cat` is working fine with the existing data for avian-flu. The difference is behavior is only in edge cases.

Reply:
Thanks! I'll happily switch to seqkit now. I'm not sure when this (or any of my avian-flu) PR(s) will be merged. They're mostly prototypes / experiments around https://github.com/nextstrain/pathogen-repo-guide/issues/74. However of all my PRs this is the easiest to rebase onto `master` and actually release, so it's plausible to do so before `augur merge --sequences` is released _if_ anyone wants it and/or we feel it's the correct approach to solving https://github.com/nextstrain/pathogen-repo-guide/issues/72. 

P.S. Do you have a timeframe for `augur merge --sequences`?

Reply:
No clear timeframe. It's a big rewrite of the `augur merge` code which needs review and testing.

Actually, I think it would be useful to test on real data such as here if you plan to not merge this soon. I can easily create a custom Docker image that has `augur merge --sequences`, would you be comfortable working from that? I could help rewrite the Snakemake rules to merge both metadata and sequences in the same `augur merge` command.

Reply:
I don't think anything beyond `seqkit` is crucial for avian-flu, but if it would be valuable as a testing platform please go for it! Could you create a branch off this one for testing this specifically?

Reply:
@jameshadfield thanks! I've added pathogen repo testing to the checklist for https://github.com/nextstrain/augur/pull/1601. Considering either this repo or mpox, which is already using an ad-hoc script to merge both metadata and sequences. I'll update here if/when I decide to test on this repo.